src/Distribution/Server/Features/AnalyticsPixels/State.hs

```diff
@@ -0,0 +1,109 @@
+{-# LANGUAGE DeriveDataTypeable, GeneralizedNewtypeDeriving,
+             TypeFamilies, TemplateHaskell #-}
+
+module Distribution.Server.Features.AnalyticsPixels.State 
+    ( AnalyticsPixel(..)
+    , AnalyticsPixelsState(..)
+    , initialAnalyticsPixelsState
+
+    -- * State queries and updates
+    , AnalyticsPixelsForPackage(..)
+    , AddPackageAnalyticsPixel(..)
+    , RemovePackageAnalyticsPixel(..)
+    , GetAnalyticsPixelsState(..)
+    , ReplaceAnalyticsPixelsState(..)
+    ) where
+
+import Distribution.Package (PackageName)
+
+import Distribution.Server.Framework.MemSize (MemSize)
+import Distribution.Server.Users.State ()
+
+import Data.Text (Text)
+import Data.Typeable (Typeable)
+import Data.Map (Map)
+import qualified Data.Map.Strict as Map
+import Data.Acid     (Query, Update, makeAcidic)
+import Data.SafeCopy (base, deriveSafeCopy)
+import Data.Set (Set)
+import qualified Data.Set as Set
+
+import Control.DeepSeq (NFData)
+import qualified Control.Monad.State as State
+import Control.Monad.Reader.Class (ask, asks)
+
+newtype AnalyticsPixel = AnalyticsPixel
+    {
+        analyticsPixelUrl :: Text
+    }
+    deriving (Show, Eq, Ord, NFData, Typeable, MemSize)
+
+newtype AnalyticsPixelsState = AnalyticsPixelsState
+    {
+        analyticsPixels :: Map PackageName (Set AnalyticsPixel)
+    }
+  deriving (Show, Eq, NFData, Typeable, MemSize)
+
+-- SafeCopy instances
+$(deriveSafeCopy 0 'base ''AnalyticsPixel)
```

Comment:
GHC 9.x's TemplateHaskell seems to process the source top-to-bottom, so the TH invocations need to be put in the right locations.  See the reorderings I performed in 9d4a4900584a87f85f3ca72a8fc907dcfbc1e21b when migrating `hackage-server` to GHC 9.0.

Reply:
Thanks for the pointer! Will take care of it

Reply:
Done!